### PR TITLE
ci: add an action to build each feature of the contained-shim-wasm crate

### PR DESCRIPTION
**Pull request by Mossaka**
This commits adds a new action job to build each feature of the `containerd-shim-wasm` crate. It adds a new script called build.sh which is able to find all the features for a given crate path, and run `cargo build -p <package-name> --features <feature>`.

### CODE REVIEW

File: scripts/build.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+PACKAGE="containerd-shim-wasm"
+
+# Check if FEATURES_FILE exists
+if [[ ! -f "$FEATURES_FILE" ]]; then
+  echo "File $FEATURES_FILE not found."
+  exit 1
+fi
+
+# captures [features] section, filters out [features], splits on =, trims whitespace
+FEATURES=$(sed -n '/\[features\]/,/^\[.*\]/p' "$FEATURES_FILE" | grep -v '^\[' | cut -d '=' -f 1 | tr -d ' ')
```
Comment (cpuguy83) [inline scripts/build.sh:32]: Have you see `cargo-whatfeatures`?

```console
$ cargo whatfeatures -t --json ./crates/cargo whatfeatures -t --json ./crates/containerd-shim-wasmedge/ | jq .members[0].features
{
  "default": [
    "standalone",
    "static"
  ],
  "standalone": [
    "wasmedge-sdk/standalone"
  ],
  "static": [
    "wasmedge-sdk/static"
  ],
  "wasi_nn": [
    "wasmedge-sdk/wasi_nn"
  ]
}
```

There is also https://github.com/frewsxcv/cargo-all-features specifically for building/testing all feature combos.

Comment (Mossaka) [reply scripts/build.sh:32]: cargo-all-features looks pretty awesome

File: crates/containerd-shim-wasm/Cargo.toml [hunk 2]
```diff
@@ -53,3 +53,21 @@ systemd = ["libcontainer/systemd"]
 cgroupsv2 = ["libcontainer/v2"]
 cgroupsv1 = ["libcontainer/v1"]
 cgroupsv2_devices = ["libcontainer/cgroupsv2_devices"]
+
+[package.metadata.cargo-all-features]
+# Features "foo" and "bar" are incompatible, so skip permutations including them
```
Comment (jsturtevant) [inline crates/containerd-shim-wasm/Cargo.toml:58]: copy pasta?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -47,6 +47,29 @@ jobs:
           echo "WASMEDGE_INCLUDE_DIR=C:\Program Files\WasmEdge\include" >> $env:GITHUB_ENV
       - name: Run checks
         run: make check
+
+  build-containerd-shim-wasm:
```
Comment (jsturtevant) [inline .github/workflows/ci.yml:51]:
```suggestion
  build-all-features-containerd-shim-wasm:
```

or something like that to distinguish it from others?

File: crates/containerd-shim-wasm/Cargo.toml (hunk 2 above)
Comment (jsturtevant) [reply crates/containerd-shim-wasm/Cargo.toml:58]: maybe a note about why `["cgroupsv2", "cgroupsv2_devices"]` etc is a combination we want to skip?

Comment (Mossaka) [reply crates/containerd-shim-wasm/Cargo.toml:58]: Ah yes that's a copy and paste issue

Comment (Mossaka) [reply crates/containerd-shim-wasm/Cargo.toml:58]: When I ran with these two features on, I got the following build error

```bash
  Compiling containerd-shim-wasm v0.2.1 (/home/mossaka/developer/runwasi/crates/containerd-shim-wasm)
error[E0308]: mismatched types
    --> /home/mossaka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libcgroups-0.1.0/src/v2/devices/bpf.rs:58:17
     |
52   |             bpf_prog_load(
     |             ------------- arguments to this function are incorrect
...
58   |                 &opts,
     |                 ^^^^^ types differ in mutability
     |
     = note: expected raw pointer `*mut bpf_prog_load_opts`
                  found reference `&bpf_prog_load_opts`
note: function defined here
    --> /home/mossaka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libbpf-sys-1.2.1+v1.2.0/src/bindings.rs:4427:12
     |
4427 |     pub fn bpf_prog_load(
     |            ^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `libcgroups` (lib) due to previous error
```

Comment (jprendes) [reply crates/containerd-shim-wasm/Cargo.toml]: @utam0k , I believe this needs to be fixed upstream.

Comment (jprendes) [reply crates/containerd-shim-wasm/Cargo.toml]: @Mossaka , I suggest you ignore the `cgroupsv2_devices` feature for now. It seems to be broken upstream, and IIUC it's a no-op without enabling `cgroupsv2`.

Comment (Mossaka) [reply crates/containerd-shim-wasm/Cargo.toml]: Removed `cgroupsv2_devices` feature

Comment (utam0k) [reply crates/containerd-shim-wasm/Cargo.toml]: @jprendes Thanks for your report 👀 
https://github.com/containers/youki/issues/new

Comment (utam0k) [reply crates/containerd-shim-wasm/Cargo.toml]: @jprendes @Mossaka I can't reproduce this error on my local. May I ask you to write how to reproduce this issue?
https://github.com/containers/youki/issues/2305

Comment (jprendes) [reply crates/containerd-shim-wasm/Cargo.toml]: I haven't tried reproduceding this locally, but just a thought. Are you using libcontainer from the `0.1` tag?

Comment (utam0k) [reply crates/containerd-shim-wasm/Cargo.toml]: hmm... I gave it a try to build it with `v0.1.0`, but I didn't get the error 🤔

**Comment by Mossaka** [comment]
PTAL 🙏 @jsturtevant @jprendes

**Review by jsturtevant — APPROVED** [review]
LGTM

**Comment by jsturtevant** [comment]
needs a rebase and may want to trigger CI again since Alot has changed...

**Comment by Mossaka** [comment]
Rebased

File: crates/containerd-shim-wasm/Cargo.toml (hunk 2 above)
Comment (YJDoc2) [reply crates/containerd-shim-wasm/Cargo.toml]: Hey, I cloned this PR added the `cgroupsv2_device` back in, and ran the `cargo build-all-features` and it completed without error as well

**Comment by cpuguy83** [comment]
Looks like this has a conflict again.

**Comment by Mossaka** [comment]
It looks like we no longer have features in `containerd-shim-wasm` crate after #314 merges.

**Pull request closed without merging**